src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs

```diff
@@ -43,41 +65,67 @@ public sealed override async Task ProvideCompletionsAsync(CompletionContext cont
+            var syntaxTree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var syntaxKinds = document.GetRequiredLanguageService<ISyntaxKindsService>();
 
-            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var declaration = GetAsyncSupportingDeclaration(root.FindToken(item.Span.Start));
-            if (declaration is null)
+            if ((completionChangeEdit & CompletionChangeEdit.MakeContainerAsync) != 0)
             {
-                // IsComplexTextEdit should only be true when GetAsyncSupportingDeclaration returns non-null.
-                // This is ensured by the ShouldMakeContainerAsync overrides.
-                Debug.Assert(false, "Expected non-null value for declaration.");
-                return await base.GetChangeAsync(document, item, commitKey, cancellationToken).ConfigureAwait(false);
+                var root = await syntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+                var declaration = GetAsyncSupportingDeclaration(root.FindToken(item.Span.Start));
+                if (declaration is null)
+                {
+                    // IsComplexTextEdit should only be true when GetAsyncSupportingDeclaration returns non-null.
+                    // This is ensured by the ShouldMakeContainerAsync overrides.
+                    Debug.Assert(false, "Expected non-null value for declaration.");
+                    return await base.GetChangeAsync(document, item, commitKey, cancellationToken).ConfigureAwait(false);
+                }
+
+                builder.Add(new TextChange(new TextSpan(GetSpanStart(declaration), 0), syntaxFacts.GetText(syntaxKinds.AsyncKeyword) + " "));
             }
 
-            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
-            var syntaxKinds = document.GetRequiredLanguageService<ISyntaxKindsService>();
+            if ((completionChangeEdit & CompletionChangeEdit.AddAwaitAtCursor) != 0)
+            {
+                builder.Add(new TextChange(item.Span, item.DisplayText));
+            }
 
-            using var _ = ArrayBuilder<TextChange>.GetInstance(out var builder);
-            builder.Add(new TextChange(new TextSpan(GetSpanStart(declaration), 0), syntaxFacts.GetText(syntaxKinds.AsyncKeyword) + " "));
-            builder.Add(new TextChange(item.Span, item.DisplayText));
+            if ((completionChangeEdit & CompletionChangeEdit.AddAwaitBeforeDotExpression) != 0)
+            {
+                var position = item.Span.Start;
+                var dotToken = GetDotTokenLeftOfPosition(syntaxTree, position, cancellationToken);
+                var expr = GetExpressionToPlaceAwaitInFrontOf(syntaxTree, position, cancellationToken);
+                if (expr is not null && dotToken.HasValue)
```

Comment:
expression should be non-null right? and dotToken shoudl haveValue right?  otherwise, something in the system went very very wrong, right?  so in this case, i would put Contract calls that these succeed.